// @expo/devcert/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uninstall = void 0;
exports.certificateFor = certificateFor;
exports.hasCertificateFor = hasCertificateFor;
exports.configuredDomains = configuredDomains;
exports.removeDomain = removeDomain;
const fs_1 = require("fs");
const debug_1 = __importDefault(require("debug"));
const constants_1 = require("./constants");
const platforms_1 = __importDefault(require("./platforms"));
const utils_1 = require("./utils");
const certificate_authority_1 = __importStar(require("./certificate-authority"));
Object.defineProperty(exports, "uninstall", { enumerable: true, get: function () { return certificate_authority_1.uninstall; } });
const certificates_1 = __importDefault(require("./certificates"));
const user_interface_1 = __importDefault(require("./user-interface"));
const debug = (0, debug_1.default)('devcert');
/**
 * Request an SSL certificate for the given app name signed by the devcert root
 * certificate authority. If devcert has previously generated a certificate for
 * that app name on this machine, it will reuse that certificate.
 *
 * If this is the first time devcert is being run on this machine, it will
 * generate and attempt to install a root certificate authority.
 *
 * Returns a promise that resolves with { key, cert }, where `key` and `cert`
 * are Buffers with the contents of the certificate private key and certificate
 * file, respectively
 *
 * If `options.getCaBuffer` is true, return value will include the ca certificate data
 * as { ca: Buffer }
 *
 * If `options.getCaPath` is true, return value will include the ca certificate path
 * as { caPath: string }
 */
async function certificateFor(domain, options = {}) {
    if (constants_1.VALID_IP.test(domain)) {
        throw new Error('IP addresses are not supported currently');
    }
    if (!constants_1.VALID_DOMAIN.test(domain)) {
        throw new Error(`"${domain}" is not a valid domain name.`);
    }
    debug(`Certificate requested for ${domain}. Skipping certutil install: ${Boolean(options.skipCertutilInstall)}. Skipping hosts file: ${Boolean(options.skipHostsFile)}`);
    if (options.ui) {
        Object.assign(user_interface_1.default, options.ui);
    }
    if (!constants_1.isMac && !constants_1.isLinux && !constants_1.isWindows) {
        throw new Error(`Platform not supported: "${process.platform}"`);
    }
    if (!(0, utils_1.commandExists)('openssl')) {
        throw new Error('OpenSSL not found: OpenSSL is required to generate SSL certificates - make sure it is installed and available in your PATH');
    }
    let domainKeyPath = (0, constants_1.pathForDomain)(domain, `private-key.key`);
    let domainCertPath = (0, constants_1.pathForDomain)(domain, `certificate.crt`);
    if (!(0, fs_1.existsSync)(constants_1.rootCAKeyPath)) {
        debug('Root CA is not installed yet, so it must be our first run. Installing root CA ...');
        await (0, certificate_authority_1.default)(options);
    }
    else if (options.getCaBuffer || options.getCaPath) {
        debug('Root CA is not readable, but it probably is because an earlier version of devcert locked it. Trying to fix...');
        await (0, certificate_authority_1.ensureCACertReadable)(options);
    }
    if (!(0, fs_1.existsSync)((0, constants_1.pathForDomain)(domain, `certificate.crt`))) {
        debug(`Can't find certificate file for ${domain}, so it must be the first request for ${domain}. Generating and caching ...`);
        await (0, certificates_1.default)(domain);
    }
    if (!options.skipHostsFile) {
        await platforms_1.default.addDomainToHostFileIfMissing(domain);
    }
    debug(`Returning domain certificate`);
    const ret = {
        key: (0, fs_1.readFileSync)(domainKeyPath),
        cert: (0, fs_1.readFileSync)(domainCertPath)
    };
    if (options.getCaBuffer)
        ret.ca = (0, fs_1.readFileSync)(constants_1.rootCACertPath);
    if (options.getCaPath)
        ret.caPath = constants_1.rootCACertPath;
    return ret;
}
function hasCertificateFor(domain) {
    return (0, fs_1.existsSync)((0, constants_1.pathForDomain)(domain, `certificate.crt`));
}
function configuredDomains() {
    return (0, fs_1.readdirSync)(constants_1.domainsDir);
}
function removeDomain(domain) {
    return (0, fs_1.rmSync)((0, constants_1.pathForDomain)(domain), { force: true, recursive: true });
}
