// @expo/devcert/dist/user-interface.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_readline_1 = __importDefault(require("node:readline"));
const utils_1 = require("./utils");
async function passwordPrompt(prompt) {
    const input = node_readline_1.default.createInterface({
        input: process.stdin,
        output: process.stdout,
    });
    return new Promise((resolve, reject) => {
        input.on('SIGINT', () => {
            reject(new Error('SIGINT'));
        });
        input.question(prompt, (answer) => {
            node_readline_1.default.moveCursor(process.stdout, 0, -1);
            node_readline_1.default.clearLine(process.stdout, 0);
            input.write(prompt + answer.replace(/./g, '*') + '\n');
            input.close();
            resolve(answer);
        });
    });
}
const DefaultUI = {
    async getWindowsEncryptionPassword() {
        return await passwordPrompt('devcert password (http://bit.ly/devcert-what-password?):');
    },
    async warnChromeOnLinuxWithoutCertutil() {
        console.warn(`
      WARNING: It looks like you have Chrome installed, but you specified
      'skipCertutilInstall: true'. Unfortunately, without installing
      certutil, it's impossible get Chrome to trust devcert's certificates
      The certificates will work, but Chrome will continue to warn you that
      they are untrusted.
    `);
    },
    async closeFirefoxBeforeContinuing() {
        console.log('Please close Firefox before continuing');
    },
    async startFirefoxWizard(certificateHost) {
        console.log(`
      devcert was unable to automatically configure Firefox. You'll need to
      complete this process manually. Don't worry though - Firefox will walk
      you through it.

      When you're ready, hit any key to continue. Firefox will launch and
      display a wizard to walk you through how to trust the devcert
      certificate. When you are finished, come back here and we'll finish up.

      (If Firefox doesn't start, go ahead and start it and navigate to
      ${certificateHost} in a new tab.)

      If you are curious about why all this is necessary, check out
      https://github.com/davewasmer/devcert#how-it-works

      <Press any key to launch Firefox wizard>
    `);
        await (0, utils_1.waitForUser)();
    },
    async firefoxWizardPromptPage(certificateURL) {
        return `
      <html>
        <head>
          <meta http-equiv="refresh" content="0; url=${certificateURL}" />
        </head>
      </html>
    `;
    },
    async waitForFirefoxWizard() {
        console.log(`
      Launching Firefox ...

      Great! Once you've finished the Firefox wizard for adding the devcert
      certificate, just hit any key here again and we'll wrap up.

      <Press any key to continue>
    `);
        await (0, utils_1.waitForUser)();
    }
};
exports.default = DefaultUI;
