// @expo/devcert/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openssl = openssl;
exports.run = run;
exports.sudoAppend = sudoAppend;
exports.waitForUser = waitForUser;
exports.reportableError = reportableError;
exports.mktmp = mktmp;
exports.sudo = sudo;
exports.commandExists = commandExists;
const child_process_1 = require("child_process");
const crypto_1 = require("crypto");
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const debug_1 = __importDefault(require("debug"));
const path_1 = __importDefault(require("path"));
const sudo_prompt_1 = __importDefault(require("@expo/sudo-prompt"));
const constants_1 = require("./constants");
const debug = (0, debug_1.default)('devcert:util');
function openssl(args) {
    return run('openssl', args, {
        stdio: 'pipe',
        env: Object.assign({
            RANDFILE: path_1.default.join((0, constants_1.configPath)('.rnd'))
        }, process.env)
    });
}
function run(cmd, args, options = {}) {
    debug(`execFileSync: \`${cmd} ${args.join(' ')}\``);
    return (0, child_process_1.execFileSync)(cmd, args, options);
}
function sudoAppend(file, input) {
    run('sudo', ['tee', '-a', file], {
        input
    });
}
function waitForUser() {
    return new Promise((resolve) => {
        process.stdin.resume();
        process.stdin.on('data', resolve);
    });
}
function reportableError(message) {
    return new Error(`${message} | This is a bug in devcert, please report the issue at https://github.com/davewasmer/devcert/issues`);
}
function mktmp() {
    const random = (0, crypto_1.randomBytes)(6).toString('hex');
    const tmppath = path_1.default.join(os_1.default.tmpdir(), `tmp-${process.pid}${random}`);
    fs_1.default.closeSync(fs_1.default.openSync(tmppath, 'w'));
    return tmppath;
}
function sudo(cmd) {
    return new Promise((resolve, reject) => {
        sudo_prompt_1.default.exec(cmd, { name: 'devcert' }, (err, stdout, stderr) => {
            let error = err || (typeof stderr === 'string' && stderr.trim().length > 0 && new Error(stderr));
            error ? reject(error) : resolve(stdout);
        });
    });
}
const _commands = {};
function commandExists(command) {
    if (_commands[command] !== undefined) {
        return _commands[command];
    }
    const paths = process.env[constants_1.isWindows ? 'Path' : 'PATH'].split(path_1.default.delimiter);
    const extensions = [...(process.env.PATHEXT || '').split(path_1.default.delimiter), ''];
    for (const dir of paths) {
        for (const extension of extensions) {
            const filePath = path_1.default.join(dir, command + extension);
            try {
                fs_1.default.accessSync(filePath, fs_1.default.constants.X_OK);
                return (_commands[command] = filePath);
            }
            catch (_a) { }
        }
    }
    return (_commands[command] = null);
}
