// @expo/devcert/dist/platforms/shared.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addCertificateToNSSCertDB = addCertificateToNSSCertDB;
exports.removeCertificateFromNSSCertDB = removeCertificateFromNSSCertDB;
exports.closeFirefox = closeFirefox;
exports.openCertificateInFirefox = openCertificateInFirefox;
exports.assertNotTouchingFiles = assertNotTouchingFiles;
const path_1 = __importDefault(require("path"));
const debug_1 = __importDefault(require("debug"));
const assert_1 = __importDefault(require("assert"));
const http_1 = __importDefault(require("http"));
const fs_1 = __importDefault(require("fs"));
const utils_1 = require("../utils");
const constants_1 = require("../constants");
const user_interface_1 = __importDefault(require("../user-interface"));
const child_process_1 = require("child_process");
const debug = (0, debug_1.default)('devcert:platforms:shared');
async function* iterateNSSCertDBPaths(nssDirGlob) {
    const globIdx = nssDirGlob.indexOf('*');
    if (globIdx === -1) {
        try {
            const stat = fs_1.default.statSync(nssDirGlob);
            if (stat.isDirectory()) {
                yield nssDirGlob;
            }
        }
        catch (_error) {
            // no matching directory found
        }
    }
    else if (globIdx === nssDirGlob.length - 1) {
        const targetDir = path_1.default.dirname(nssDirGlob);
        for (const entry of await fs_1.default.promises.readdir(targetDir, { withFileTypes: true })) {
            if (entry.isDirectory()) {
                yield path_1.default.join(targetDir, entry.name);
            }
        }
    }
    else {
        throw new Error('Internal: Invalid `nssDirGlob` specified');
    }
}
async function* iterateNSSCertDBs(nssDirGlob) {
    for await (const dir of iterateNSSCertDBPaths(nssDirGlob)) {
        debug(`checking to see if ${dir} is a valid NSS database directory`);
        if (fs_1.default.existsSync(path_1.default.join(dir, 'cert8.db'))) {
            debug(`Found legacy NSS database in ${dir}, emitting...`);
            yield { dir, version: 'legacy' };
        }
        if (fs_1.default.existsSync(path_1.default.join(dir, 'cert9.db'))) {
            debug(`Found modern NSS database in ${dir}, running callback...`);
            yield { dir, version: 'modern' };
        }
    }
}
/**
 *  Given a directory or glob pattern of directories, attempt to install the
 *  CA certificate to each directory containing an NSS database.
 */
async function addCertificateToNSSCertDB(nssDirGlob, certPath, certutilPath) {
    debug(`trying to install certificate into NSS databases in ${nssDirGlob}`);
    for await (const { dir, version } of iterateNSSCertDBs(nssDirGlob)) {
        const dirArg = version === 'modern' ? `sql:${dir}` : dir;
        (0, utils_1.run)(certutilPath, ['-A', '-d', dirArg, '-t', 'C,,', '-i', certPath, '-n', 'devcert']);
    }
    debug(`finished scanning & installing certificate in NSS databases in ${nssDirGlob}`);
}
async function removeCertificateFromNSSCertDB(nssDirGlob, certPath, certutilPath) {
    debug(`trying to remove certificates from NSS databases in ${nssDirGlob}`);
    for await (const { dir, version } of iterateNSSCertDBs(nssDirGlob)) {
        const dirArg = version === 'modern' ? `sql:${dir}` : dir;
        try {
            (0, utils_1.run)(certutilPath, ['-A', '-d', dirArg, '-t', 'C,,', '-i', certPath, '-n', 'devcert']);
        }
        catch (e) {
            debug(`failed to remove ${certPath} from ${dir}, continuing. ${e.toString()}`);
        }
    }
    debug(`finished scanning & installing certificate in NSS databases in ${nssDirGlob}`);
}
/**
 *  Check to see if Firefox is still running, and if so, ask the user to close
 *  it. Poll until it's closed, then return.
 *
 * This is needed because Firefox appears to load the NSS database in-memory on
 * startup, and overwrite on exit. So we have to ask the user to quite Firefox
 * first so our changes don't get overwritten.
 */
async function closeFirefox() {
    if (isFirefoxOpen()) {
        await user_interface_1.default.closeFirefoxBeforeContinuing();
        while (isFirefoxOpen()) {
            await sleep(50);
        }
    }
}
/**
 * Check if Firefox is currently open
 */
function isFirefoxOpen() {
    // NOTE: We use some Windows-unfriendly methods here (ps) because Windows
    // never needs to check this, because it doesn't update the NSS DB
    // automaticaly.
    (0, assert_1.default)(constants_1.isMac || constants_1.isLinux, 'checkForOpenFirefox was invoked on a platform other than Mac or Linux');
    return (0, child_process_1.execSync)('ps aux').indexOf('firefox') > -1;
}
async function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
/**
 * Firefox manages it's own trust store for SSL certificates, which can be
 * managed via the certutil command (supplied by NSS tooling packages). In the
 * event that certutil is not already installed, and either can't be installed
 * (Windows) or the user doesn't want to install it (skipCertutilInstall:
 * true), it means that we can't programmatically tell Firefox to trust our
 * root CA certificate.
 *
 * There is a recourse though. When a Firefox tab is directed to a URL that
 * responds with a certificate, it will automatically prompt the user if they
 * want to add it to their trusted certificates. So if we can't automatically
 * install the certificate via certutil, we instead start a quick web server
 * and host our certificate file. Then we open the hosted cert URL in Firefox
 * to kick off the GUI flow.
 *
 * This method does all this, along with providing user prompts in the terminal
 * to walk them through this process.
 */
async function openCertificateInFirefox(firefoxPath, certPath) {
    debug('Adding devert to Firefox trust stores manually. Launching a webserver to host our certificate temporarily ...');
    let port;
    const server = http_1.default.createServer(async (req, res) => {
        let { pathname } = new URL(req.url);
        if (pathname === '/certificate') {
            res.writeHead(200, { 'Content-type': 'application/x-x509-ca-cert' });
            res.write(fs_1.default.readFileSync(certPath));
            res.end();
        }
        else {
            res.writeHead(200);
            res.write(await user_interface_1.default.firefoxWizardPromptPage(`http://localhost:${port}/certificate`));
            res.end();
        }
    });
    port = await new Promise((resolve, reject) => {
        server.on('error', reject);
        server.listen(() => {
            resolve(server.address().port);
        });
    });
    try {
        debug('Certificate server is up. Printing instructions for user and launching Firefox with hosted certificate URL');
        await user_interface_1.default.startFirefoxWizard(`http://localhost:${port}`);
        (0, utils_1.run)(firefoxPath, [`http://localhost:${port}`]);
        await user_interface_1.default.waitForFirefoxWizard();
    }
    finally {
        server.close();
    }
}
function assertNotTouchingFiles(filepath, operation) {
    if (!filepath.startsWith(constants_1.configDir) && !filepath.startsWith((0, constants_1.getLegacyConfigDir)())) {
        throw new Error(`Devcert cannot ${operation} ${filepath}; it is outside known devcert config directories!`);
    }
}
